File: aoj/src/ALDS1_5_A/main.py
ans = {}

def exhaustive_search(current, input):
    global ans
    if len(input) == 0:
        return False
    next = current + input[0]
    ans[next] = True
    exhaustive_search(next, input[1:])
    exhaustive_search(current, input[1:])
    
    

# ALDS1_5_A: 全探索
def main():
    _ = input()
    s = [int(v) for v in input().split(" ")]
    _ = input()
    q = [int(v) for v in input().split(" ")]
    global ans
    ans[0] = True
    exhaustive_search(0, s)
    for qv in q:
        if qv in ans:
            print("yes")
        else:
            print("no")

if __name__ == '__main__':
    main()

File: aoj/src/ALDS1_5_A/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const maxWeight = 10000

// 配列Aの要素を組み合わせて配列mの要素の値を作れるか全探索で調べる
// n個の配列Aとq個の配列M
func main() {
	stdin := bufio.NewScanner(os.Stdin)
	stdin.Scan() // skip n
	a := scanIntArray(stdin)
	stdin.Scan() // skip q
	m := scanIntArray(stdin)
	for _, mi := range m {
		if exhaustiveSearch(a, mi) {
			fmt.Println("yes")
		} else {
			fmt.Println("no")
		}
	}
}

func exhaustiveSearch(a []int, mi int) bool {
	if len(a) == 0 {
		return false
	}
	if a[0] == mi {
		return true
	}

	if a[0] < mi {
		return exhaustiveSearch(a[1:], mi-a[0]) || exhaustiveSearch(a[1:], mi)
	}

	return exhaustiveSearch(a[1:], mi)
}

func scanInt(sc *bufio.Scanner) int {
	sc.Scan()
	t := sc.Text()
	i, err := strconv.Atoi(t)
	if err != nil {
		panic(err)
	}
	return i
}

func scanIntArray(sc *bufio.Scanner) []int {
	sc.Scan()
	t := sc.Text()
	ss := strings.Split(t, " ")
	is := []int{}
	for _, s := range ss {
		i, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		is = append(is, i)
	}
	return is
}
